pkg/api/login_oauth.go

```diff
@@ -256,3 +274,31 @@ func hashStatecode(code, seed string) string {
 	hashBytes := sha256.Sum256([]byte(code + setting.SecretKey + seed))
 	return hex.EncodeToString(hashBytes[:])
 }
+
+func (hs *HTTPServer) handleOAuthLoginError(ctx *models.ReqContext, action string, status int, errMsg string, err error) {
```

Comment:
I find it hard to follow when there are more than ~3 arguments to a function, this structure could be replaced by something like:

```suggestion
type loginError struct {
	action string
	httpStatus int
	publicMessage string
	err error
}

func (hs *HTTPServer) handleOAuthLoginError(ctx *models.ReqContext, loginError loginError) {
```